hawc/apps/assessment/views.py

```diff
@@ -351,25 +356,34 @@ def get_success_url(self):
 
 
 # Dataset views
-class DatasetCreate(BaseCreate):
+class DatasetPublishedMixin:
+    def get_object(self, **kwargs):
+        obj = super().get_object(**kwargs)
+        if obj.assessment.user_can_edit_object(self.request.user) or obj.published:
+            return obj
+        else:
+            raise PermissionDenied
+
+
+class DatasetCreate(DatasetPublishedMixin, BaseCreate):
```

Comment:
we dont need this mixin for any of the create/update/delete views; the base classes check permissions that users can edit content in an assessment. I think you can just move it to the Read view